File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        
        n = len(nums)
        res = [1] * n
        
        runningprefix = 1
        for i in range(0, n):
            res[i] *= runningprefix
            runningprefix *= nums[i]
        
        runningsuffix = 1
        for i in range(n-1,-1,-1):
            res[i] *= runningsuffix
            runningsuffix *=nums[i]
        
        return res

File: 238-product-of-array-except-self/product-of-array-except-self.java
public class Solution {
public int[] productExceptSelf(int[] nums) {
    int n = nums.length;
    int[] res = new int[n];
    
    int runningprefix = 1;
    for(int i = 0; i < n; i++){
        res[i] = runningprefix;
        runningprefix*= nums[i];
    }
    int runningsufix = 1;
    for(int i = n -1; i >= 0; i--){
        res[i] *= runningsufix;
        runningsufix *= nums[i];
    }
    return res;
    
}
}